File: client/src/game-portal/src/game/maps/unitDefs.ts
import type { JsonObject, UnitCapability } from '../network/protocol'
import type { UnitDirection } from '../rendering/unitSprites'

// Visual footprint of a unit, in canvas pixels relative to its (x, y) anchor.
// Used to anchor the sprite's feet, size the selection ring, place overhead UI,
// and derive hit-test rects. Sprites are centered horizontally on unit.x; top
// and bottom are signed offsets from unit.y (top is typically negative).
export type UnitBounds = {
  halfWidth: number
  top: number
  bottom: number
  // Per-unit nudge for the selection ring (and its hover/inspected/drag-select
  // siblings + perk auras), independent of where the sprite is anchored.
  // Positive Y pushes the ring down, negative Y pulls it up.
  ringOffsetX?: number
  ringOffsetY?: number
}

// Per-unit ground-shadow tuning. Every field is optional; absent fields fall
// back to defaults derived from the unit's bounds (see resolveUnitShadow in
// unitShadow.ts), so a unit with no `shadow` block still gets a proportional
// blob shadow. Passed through the catalog as-is; the server never reads it.
export type UnitShadow = {
  /** Default true. Set false to draw no shadow for this unit. */
  enabled?: boolean
  /** Horizontal radius in px. Default: bounds.halfWidth * 0.85. */
  radiusX?: number
  /** Vertical radius in px. Default: resolved radiusX * 0.4. */
  radiusY?: number
  /** Peak alpha at the center, 0..1. Default: 0.35. */
  opacity?: number
  /** Px nudge from the feet anchor. Default: 0. */
  offsetX?: number
  /** Px nudge from the feet anchor (positive = down). Default: 0. */
  offsetY?: number
}

export type UnitAttackVisual = {
  kind?: 'melee' | 'projectile'
  originX?: number
  originY?: number
  effectLength?: number
}

export type ResolvedUnitAttackVisual = {
  kind: 'melee' | 'projectile'
  originX: number
  originY: number
  effectLength: number
}

// Screen-space offset from unit.x/unit.y (the same lift the renderer applies)
// where a unit's projectiles/spells/beams visually leave its body.
export type UnitOriginPoint = { x: number; y: number }

// Per-facing attack origin authoring. Both fields are optional so an
// unauthored unit produces an empty (or absent) block — see
// getResolvedUnitAttackOrigin, which returns null in that case so the
// renderer falls through to today's exact geometry.
export type UnitAttackOrigin = {
  default?: UnitOriginPoint
  byFacing?: Partial<Record<UnitDirection, UnitOriginPoint>>
}

/** Default categorisation of a unit type for the map editor brushing flow.
 *  Decoupled from runtime ownership — the placed-unit `playerSlot` decides who
 *  controls the unit in-game; faction just filters which types appear under
 *  each brush category.
 *
 *  Open-ended string by design: factions are derived from the server catalog
 *  at runtime (one bucket per `catalog/units/<faction>/` directory), so adding
 *  a new faction folder + unit picks it up automatically without a type edit. */
export type UnitFaction = string

/** Target classes a unit's attacks are valid against. Drives anti-air / ground-only filtering on both server and client. */
export type UnitTargetClass = 'ground' | 'flyer'

export type UnitDef = {
  type: string
  name: string
  faction: UnitFaction
  archetype?: string
  hp: number
  damage: number
  attackRange: number
  attackSpeed: number
  /** Base pixels-per-second pathing speed (pre-rank/path/perk modifiers). */
  moveSpeed: number
  goldGatherAmount?: number
  woodGatherAmount?: number
  resourceCost: Record<string, number>
  meatCost: number
  spawnSeconds: number
  capabilities: UnitCapability[]
  trainLabel: string
  metadata?: JsonObject
  /** Server-only: name of the AI combat profile to use. Ignored by the client. */
  combatProfile?: string
  attackVisual?: UnitAttackVisual
  /** Per-facing point where projectiles/spells/beams leave this unit's body,
   *  as a screen-space offset from unit.x/unit.y. Absent ⇒ getResolvedUnitAttackOrigin
   *  returns null and the renderer keeps its existing geometry unchanged. */
  attackOrigin?: UnitAttackOrigin
  bounds?: UnitBounds
  /** Optional ground-shadow tuning. Absent ⇒ defaults derived from bounds. */
  shadow?: UnitShadow
  /** Airborne unit. Renders above ground units (with shadow/elevation) and is only hit by attackers whose targetableTypes include "flyer". */
  flyer?: boolean
  /** Target classes this unit's attacks can hit. When absent the server derives a default at spawn (projectile attacks → ground+flyer, otherwise ground only). */
  targetableTypes?: UnitTargetClass[]
  /** Building types the player must own fully-built before this unit
   *  can be trained. Server is authoritative; client uses this only to
   *  render the requirement tooltip on locked train actions. */
  requiresBuildings?: string[]
}

export let UNIT_DEFS: UnitDef[] = []

export let UNIT_DEF_MAP = new Map<string, UnitDef>()

export function initUnitDefs(defs: UnitDef[]): void {
  UNIT_DEFS = defs
  UNIT_DEF_MAP = new Map(defs.map((def) => [def.type, def]))
}

// Fallback used when a unit def lacks `bounds` — roughly the pre-refactor
// 14px feet-radius body so hit-testing and ring sizing still work.
export const DEFAULT_UNIT_BOUNDS: UnitBounds = {
  halfWidth: 14,
  top: -26,
  bottom: 2,
}

// Merge the authored bounds OVER the defaults rather than replacing them, so a
// PARTIAL bounds block (e.g. the anchors overlay authoring only ringOffsetX/Y,
// with no halfWidth/top/bottom) still yields a complete UnitBounds. Replacing
// would leave halfWidth/bottom `undefined`, which anchors the sprite at NaN and
// makes the unit render invisibly.
export function getUnitBounds(def: UnitDef | undefined | null): UnitBounds {
  return def?.bounds ? { ...DEFAULT_UNIT_BOUNDS, ...def.bounds } : DEFAULT_UNIT_BOUNDS
}

// Per-path bounds override: path id (e.g. "marksman") → bounds. Populated
// from the /catalog/units `paths` array. Path-promoted units use these
// before falling back to the base UnitDef bounds — necessary because path
// variants ship their own sprites with different pixel sizes.
export let PATH_BOUNDS_MAP: Map<string, UnitBounds> = new Map()

export function initPathBounds(entries: Array<{ path: string; bounds: UnitBounds }>): void {
  PATH_BOUNDS_MAP = new Map(entries.map((e) => [e.path, e.bounds]))
}

// Per-path ground-shadow override: path id → shadow config. Populated from the
// /catalog/units `paths` array's shadow field. Mirrors PATH_BOUNDS_MAP /
// PATH_ATTACK_ORIGIN_MAP — a path variant may want its own blob shadow, distinct
// from the base unit's (authored via UnitSpritePreview's anchors overlay).
export let PATH_SHADOW_MAP: Map<string, UnitShadow> = new Map()

export function initPathShadow(entries: Array<{ path: string; shadow?: UnitShadow | null }>): void {
  PATH_SHADOW_MAP = new Map(
    entries
      .filter((e): e is { path: string; shadow: UnitShadow } => !!e.shadow)
      .map((e) => [e.path, e.shadow]),
  )
}

// Resolves the shadow config for a rendered unit instance, checking the PATH's
// authored shadow before the base unit def's — mirrors getUnitBoundsFor's
// path-then-type precedence. Returns undefined when neither authors one, so the
// caller falls through to resolveUnitShadow's bounds-derived default blob.
export function getUnitShadowFor(args: {
  path?: string | null
  unitType?: string | null
}): UnitShadow | undefined {
  if (args.path && args.path !== 'none') {
    const s = PATH_SHADOW_MAP.get(args.path)
    if (s) return s
  }
  const def = args.unitType ? UNIT_DEF_MAP.get(args.unitType) : undefined
  return def?.shadow
}

// Catalog topology: unit type → its promotion path ids, mirroring the
// server directory layout under catalog/units/<faction>/<unit>/paths/.
// Populated from the /catalog/units `pathsByUnit` field. Consumers (e.g.
// DebugSpawnPanel) should derive their unit→path UI from this map instead
// of duplicating the layout in code.
export let PATHS_BY_UNIT_TYPE_MAP: Map<string, string[]> = new Map()

export function initPathsByUnitType(byUnit: Record<string, string[]>): void {
  PATHS_BY_UNIT_TYPE_MAP = new Map(Object.entries(byUnit))
}

// Resolves bounds for a unit instance, checking path before unitType. Mirrors
// the path-then-type lookup in getUnitSpriteSet so a path-promoted unit's
// selection ring tracks its own sprite, not the base unit's.
export function getUnitBoundsFor(args: {
  path?: string | null
  unitType?: string | null
}): UnitBounds {
  const def = args.unitType ? UNIT_DEF_MAP.get(args.unitType) : undefined
  const base = getUnitBounds(def)
  if (args.path && args.path !== 'none') {
    const pathBounds = PATH_BOUNDS_MAP.get(args.path)
    // Merge the path override OVER the base unit's bounds — a path may author a
    // COMPLETE bounds (its sprite is a different size) or just a PARTIAL tweak
    // (e.g. only the ring offset via the anchors overlay). Either way it inherits
    // any field it doesn't set from the base unit instead of nulling it out.
    if (pathBounds) return { ...base, ...pathBounds }
  }
  return base
}

export function getResolvedUnitAttackVisual(
  def: UnitDef | undefined | null,
): ResolvedUnitAttackVisual {
  const bounds = getUnitBounds(def)
  const inferredKind = (def?.attackRange ?? 0) > 80 ? 'projectile' : 'melee'
  const kind = def?.attackVisual?.kind ?? inferredKind

  return {
    kind,
    originX: Math.round(def?.attackVisual?.originX ?? 0),
    originY: Math.round(def?.attackVisual?.originY ?? 0),
    effectLength: Math.max(
      4,
      Math.round(
        def?.attackVisual?.effectLength
          ?? (kind === 'projectile'
            ? Math.max(18, Math.round(bounds.bottom * 1.2))
            : 10),
      ),
    ),
  }
}

// Shared "byFacing wins over default" resolution for a single authored
// attackOrigin block. Extracted so the unit-only resolver (Phase 4, below)
// and the path-aware resolver (further below) share exactly one
// implementation of that precedence instead of two copies that could drift.
function resolveOriginFromBlock(
  ao: UnitAttackOrigin | null | undefined,
  facing: UnitDirection | undefined,
): UnitOriginPoint | null {
  if (!ao) return null
  const pick = (facing && ao.byFacing?.[facing]) || ao.default
  if (!pick) return null
  return { x: Math.round(pick.x), y: Math.round(pick.y) }
}

// Resolves the authored attack origin for a unit's current facing. Returns
// null when the unit has no attackOrigin authored at all (or it authors
// neither `default` nor a matching `byFacing` entry) — callers MUST treat
// null as "fall back to today's existing geometry", never as (0, 0).
export function getResolvedUnitAttackOrigin(
  def: UnitDef | null | undefined,
  facing: UnitDirection | undefined,
): UnitOriginPoint | null {
  return resolveOriginFromBlock(def?.attackOrigin, facing)
}

// Per-path attack-origin override: path id (e.g. "marksman") → its authored
// {default?, byFacing?} block. Populated from the /catalog/units `paths`
// array's (now unioned) attackOrigin field. Mirrors PATH_BOUNDS_MAP exactly
// — same map/init/lookup shape, same reason (path variants author their own
// projectile launch points, distinct from the base unit's).
export let PATH_ATTACK_ORIGIN_MAP: Map<string, UnitAttackOrigin> = new Map()

export function initPathAttackOrigin(
  entries: Array<{ path: string; attackOrigin?: UnitAttackOrigin | null }>,
): void {
  PATH_ATTACK_ORIGIN_MAP = new Map(
    entries
      .filter((e): e is { path: string; attackOrigin: UnitAttackOrigin } => !!e.attackOrigin)
      .map((e) => [e.path, e.attackOrigin]),
  )
}

// Resolves the attack origin for a rendered unit instance, checking the
// PATH's authored origin before the base unit def's — mirrors
// getUnitBoundsFor's path-then-type precedence EXACTLY, including the
// 'none' sentinel check: if the unit is on a path AND that path authored
// ANY attackOrigin block, the path's block is used exclusively (byFacing >
// default WITHIN that block) and the base unit def's attackOrigin is never
// consulted — same "path presence switches the whole source" rule bounds
// uses, not a per-field merge. Only when the path has no attackOrigin block
// at all (or there is no path) does resolution fall through to the base
// unit def, then to null, so the renderer's existing geometric fallback is
// completely unchanged for any unit/path with no authored origin anywhere.
export function getResolvedAttackOriginFor(
  args: { path?: string | null; unitType?: string | null },
  facing: UnitDirection | undefined,
): UnitOriginPoint | null {
  if (args.path && args.path !== 'none') {
    const pathOrigin = PATH_ATTACK_ORIGIN_MAP.get(args.path)
    if (pathOrigin) return resolveOriginFromBlock(pathOrigin, facing)
  }
  const def = args.unitType ? UNIT_DEF_MAP.get(args.unitType) : undefined
  return getResolvedUnitAttackOrigin(def, facing)
}

File: server/internal/game/unit_defs.go
package game

import (
	"embed"
	"encoding/json"
	"io/fs"
	"sort"
)

// Embeds the entire per-unit catalog tree. Layout:
//
//	catalog/units/<faction>/<unit>/<unit>.json   — UnitDef for that unit (loaded here)
//	catalog/units/<faction>/<unit>/paths/*.json  — per-path stat modifiers
//	                                                (loaded by path_defs.go)
//
// Adding a new unit: create catalog/units/<faction>/<newunit>/<newunit>.json.
// The faction directory name is taken as-is (no allowlist) and must match the
// JSON's `faction` field; mismatch panics at startup. Adding a brand-new
// faction is the same operation: just place the new <faction> directory in
// catalog/units and drop a unit folder inside.
//
//go:embed catalog/units
var unitDefsFS embed.FS

// UnitDef holds the configuration for a trainable unit type.
// Client-only fields (TrainLabel, Bounds) are passed through to the API
// as-is; the server game logic never reads them.
type UnitDef struct {
	Type string `json:"type"`
	Name string `json:"name"`
	// Faction categorises the unit's default origin for map editor brushing:
	// "raider" | "neutral" | "human". Decoupled from runtime ownership — a
	// "raider" unit can still be assigned to a player slot in the editor for
	// scenarios where the player takes over a raider squad. Required.
	Faction   string `json:"faction"`
	Archetype string `json:"archetype,omitempty"`
	// NonCombat marks the unit as passive: it will not auto-acquire targets in
	// the combat AI, and only engages when the player issues an explicit
	// OrderAttackTarget (via AttackWithUnits). The unit still carries the
	// `"attack"` capability so the player's attack command is accepted.
	NonCombat bool `json:"nonCombat,omitempty"`
	HP        int  `json:"hp"`
	// Armor is the catalog base armor for this unit type. All unit catalog JSON
	// files carry an explicit "armor" field (0 when the unit has no base armor,
	// 33 for soldier). The value is used directly by applyRankModifiersLocked to
	// set unit.Armor for unpathed units, and contributes the advancement-bonus
	// delta for promoted units. Do NOT seed unit.BaseArmor from this field at
	// spawn — BaseArmor is reserved for player-upgrade-track armor only
	// (applyPlayerUpgradesAtSpawnLocked).
	Armor int `json:"armor,omitempty"`
	// SpawnExp is pre-loaded XP granted to a unit at spawn (before any rank
	// modifiers run). Zero-value default is safe; units without this field
	// spawn at 0 XP as before. Used by the "Veteran Initiates" advancement.
	SpawnExp    int     `json:"spawnExp,omitempty"`
	Damage      int     `json:"damage"`
	AttackRange float64 `json:"attackRange"`
	AttackSpeed float64 `json:"attackSpeed"`
	// SplashRadius: when > 0, every attack landing on a primary target also
	// deals the same damage to every other hostile within this radius of the
	// target's position. Direct damage only — does NOT trigger on-attack
	// perks on the splashed targets (so it doesn't chain hunters_mark,
	// savage_strikes, etc.). Friendly fire excluded.
	SplashRadius float64 `json:"splashRadius,omitempty"`
	// MoveSpeed: base pixels-per-second pathing speed. Path multipliers
	// (pathModifierTable) and perk multipliers (momentum) stack on top of it.
	MoveSpeed        float64        `json:"moveSpeed"`
	GoldGatherAmount int            `json:"goldGatherAmount,omitempty"`
	WoodGatherAmount int            `json:"woodGatherAmount,omitempty"`
	ResourceCost     map[string]int `json:"resourceCost"`
	MeatCost         int            `json:"meatCost"`
	SpawnSeconds     float64        `json:"spawnSeconds"`
	Capabilities     []string       `json:"capabilities"`
	TrainLabel       string         `json:"trainLabel,omitempty"`
	// CombatProfile picks the AI behavior profile (target scoring, detection
	// range, ranged-vs-melee, etc.) from combatProfiles in combat_ai_profiles.go.
	// When empty, the server falls back to inferCombatArchetype's hardcoded
	// mapping. Validated against combatProfiles at init; unknown names panic.
	CombatProfile string          `json:"combatProfile,omitempty"`
	AttackVisual  json.RawMessage `json:"attackVisual,omitempty"`
	// AttackType names the melee attack sound the client plays when this unit's
	// swing resolves — "swing", "stab", etc. (keyed to a file in the client's
	// audio/sfx/combat folder). Melee-only: ranged units leave this empty and
	// get their sound from the projectile they fire. A promotion path may
	// override it (see pathCatalogFile.AttackType), e.g. a soldier ("swing")
	// promoting to the vanguard path ("stab"). Purely presentational — the
	// simulation never reads it.
	AttackType string `json:"attackType,omitempty"`
	// Bounds describes the unit's visual footprint (halfWidth, top, bottom
	// offsets from unit.x/unit.y). Client uses it to anchor the sprite's
	// feet, size the selection ring, and compute hit-test rects. Passed
	// through as-is; the server game logic never reads it.
	Bounds json.RawMessage `json:"bounds,omitempty"`
	// Shadow is optional per-unit ground-shadow tuning (enabled, radiusX,
	// radiusY, opacity, offsetX, offsetY). Client-only render config; the
	// server never reads it and only passes it through. Absent ⇒ the client
	// derives a default blob shadow from Bounds.
	Shadow json.RawMessage `json:"shadow,omitempty"`

	// DominionPointDropChance is the per-kill probability that this unit type
	// drops dominion points when killed by a player. Must be in [0,1].
	// Zero means no drop. Overrides the base tuning value from gameplay_tuning.json.
	DominionPointDropChance float64 `json:"dominionPointDropChance,omitempty"`
	// DominionPointAmount is how many dominion points drop when the drop chance
	// triggers. Must be >= 0. Overrides the base tuning value.
	DominionPointAmount int `json:"dominionPointAmount,omitempty"`

	// Experience is the raw XP this unit yields when killed in "split" mode
	// (catalog/tuning/gameplay_tuning.json experience.mode). Pointer so the
	// catalog can distinguish absent (→ splitDefaultXP) from an explicit 0
	// (unit grants no XP). Ignored entirely in "classic" mode.
	Experience *int `json:"experience,omitempty"`

	// VisionRange is the base vision radius in world pixels. When 0 or absent,
	// the spawn path falls back to defaultVisionRange.
	VisionRange float64 `json:"visionRange,omitempty"`

	// Flyer marks the unit as airborne. Flyers ignore terrain and ground-unit
	// obstacles when pathing — only map bounds and other flyers constrain
	// them. They are also a distinct target class: a unit can only attack a
	// flyer if "flyer" appears in its TargetableTypes.
	Flyer bool `json:"flyer,omitempty"`

	// ── Spellcaster kit (optional; zero values = a non-caster unit) ─────────
	// MaxMana / ManaRegenRate seed the unit's mana pool at spawn (Part 3
	// mechanics). CurrentMana starts at MaxMana. Absent ⇒ 0 ⇒ no mana.
	MaxMana       int     `json:"maxMana,omitempty"`
	ManaRegenRate float64 `json:"manaRegenRate,omitempty"`
	// Projectile is the id of a ProjectileDef (Part 1) this unit's basic
	// ranged attack fires (e.g. "fire_bolt"). Empty ⇒ the default procedural
	// shot (Variant = unit type), unchanged from before. Validated at load.
	Projectile string `json:"projectile,omitempty"`
	// DamageType tags this unit's basic attack damage (Part 2). Optional
	// flavor/metadata; empty ⇒ physical. Validated at load when non-empty.
	DamageType DamageType `json:"damageType,omitempty"`
	// ProjectileScale is a render-size multiplier for this unit's projectile
	// sprite, applied client-side on top of the base projectile-sprite scale
	// (e.g. 2 ⇒ twice as large, 0.5 ⇒ half). It is per-unit, not per
	// projectile def, so two units sharing one projectile (e.g. "fire_bolt")
	// can size it differently; a promotion path may override it (see
	// pathCatalogFile.ProjectileScale). Purely visual — never read by the
	// simulation. Omitted / 0 ⇒ the client default (1×). Must be >= 0;
	// validated at load.
	ProjectileScale float64 `json:"projectileScale,omitempty"`
	// Abilities is the unit's ability id list (AbilityDef ids, Part 6), slot
	// order. Not validated at load — an ability may be authored in a later
	// part; resolution is fail-safe at use (getAbilityDef).
	Abilities []string `json:"abilities,omitempty"`

	// TargetableTypes is the set of target classes this unit's attacks are
	// valid against. Recognised entries: "ground", "flyer". When empty, the
	// default is derived at spawn time from AttackVisual.kind: a projectile
	// attack defaults to ["ground","flyer"], any other attack defaults to
	// ["ground"]. Authoring an explicit value overrides the default — e.g.
	// "anti-air only" units would author ["flyer"].
	TargetableTypes []string `json:"targetableTypes,omitempty"`

	// RequiresBuildings is the list of building types the player must own
	// fully built (Visible, not underConstruction) before this unit can
	// be trained. Empty/omitted = no requirement. Multiple entries are
	// ANDed. Validated at load time against the building catalog.
	RequiresBuildings []string `json:"requiresBuildings,omitempty"`

	// ChannelLoop, when set, defines the inclusive frame range the client
	// one-way loops through on this unit's casting sprite sheet while it is
	// channeling a beam ability (Siphon Life, etc.). Pointer-to-struct so
	// the loader can distinguish "field absent" (no channel pose authored —
	// degenerates to frame 0 hold) from "field present" (use these frames).
	// A promotion path may override this via pathCatalogFile.ChannelLoop —
	// resolution order is path > unit. Validated at load (start >= 0,
	// end >= start). Purely visual; server simulation never reads it.
	ChannelLoop *ChannelLoopRange `json:"channelLoop,omitempty"`

	// PathChances is the weighted distribution over promotion paths this unit
	// type rolls the first time it reaches Bronze rank. Keyed by path id (a
	// directory under catalog/units/<faction>/<unit>/paths/), value is a
	// RELATIVE weight: the roll normalizes by the sum, so {"trapper":1,
	// "marksman":1} is a 50/50 split and {"trapper":7,"marksman":3} is 70/30.
	// A single entry (e.g. {"arch_mage":1}) is a guaranteed promotion. Empty /
	// absent ⇒ the unit type has no promotion path (workers, raiders) and
	// stays on unitPathNone. Replaces the old hardcoded per-type switch in
	// assignUnitPathOnRankUpLocked. Validated at load: every key must be a
	// real path dir under this unit (path_defs.go init), every weight >= 0,
	// and the sum must be > 0 when the map is non-empty (loadUnitDefsByType).
	PathChances map[string]float64 `json:"pathChances,omitempty"`

	// ── Advancement-granted bonuses (not authored in unit JSON) ──────────────
	// These three fields are zero in the catalog and are only set by the Archer
	// "Master Huntsman" advancement (advancement_defs.go effect kinds
	// unitBonusArrows / unitTrapEffectMul / unitTrapRadiusMul) on a player's
	// EffectiveUnitDefs copy. They flow def → unit at spawn, so only units whose
	// effective def was modified (i.e. the owning player's archers) carry them.

	// BonusArrows is the number of extra arrows this unit fires per attack on
	// top of any split_shot perk, routed through the split-shot fan-out
	// (fireSplitShotsLocked). Zero ⇒ no bonus arrow.
	BonusArrows int `json:"-"`
	// TrapEffectBonus is an additive fraction applied to the unit's trap
	// EffectMultiplier: the trap pipeline multiplies by (1 + TrapEffectBonus),
	// so 1.0 ⇒ ×2 trap effect strength. Zero ⇒ no change.
	TrapEffectBonus float64 `json:"-"`
	// TrapRadiusBonus is the same additive fraction for the trap
	// RadiusMultiplier: (1 + TrapRadiusBonus), so 1.0 ⇒ ×2 trap radius. Zero ⇒
	// no change.
	TrapRadiusBonus float64 `json:"-"`
}

// ChannelLoopRange is an inclusive [Start, End] frame range on a unit's
// casting sprite sheet. The client one-way loops through these frames at
// the unit's normal frame cadence while the unit is channeling. start == end
// produces a single held pose; start < end produces a small loop. Defined
// at package scope so UnitDef and pathCatalogFile can share the type.
type ChannelLoopRange struct {
	Start int `json:"start"`
	End   int `json:"end"`
}

// Target class strings recognised by TargetableTypes. Kept as a small closed
// set so misspellings in JSON are caught at catalog load.
const (
	TargetClassGround = "ground"
	TargetClassFlyer  = "flyer"
)

// unitDefsByType MUST remain a var initializer (not init()) because
// maps.go's mapCatalog var initializer references it via
// `_ = unitDefsByType` to force dependency-ordered loading
// (placedUnits hydration calls getUnitDef during catalog load).
// All var initializers run before any init() function, so converting
// this to init() would cause every map's placedUnits to be silently
// dropped at startup with "unknown unitType" warnings.
//
// loadUnitDefsByType validates each unit's RequiresBuildings against
// the building catalog, which means buildingDefsByType must also be
// a var initializer; Go's dependency analysis then orders building
// defs before unit defs automatically.
var unitDefsByType = loadUnitDefsByType()

func loadUnitDefsByType() map[string]UnitDef {
	// Two-level directory layout: catalog/units/<faction>/<unit>/<unit>.json.
	// Faction directory names are accepted as-is; the unit directory name
	// must match the JSON's "type" field; the JSON's "faction" field must
	// match its parent directory. Any drift panics at startup so the catalog
	// stays coherent.
	factionEntries, err := fs.ReadDir(unitDefsFS, "catalog/units")
	if err != nil {
		panic("catalog/units: " + err.Error())
	}
	result := make(map[string]UnitDef, 16)
	for _, factionEntry := range factionEntries {
		if !factionEntry.IsDir() {
			panic("catalog/units: unexpected file at root " + factionEntry.Name() + " — top-level entries must be faction directories")
		}
		factionKey := factionEntry.Name()
		unitEntries, err := fs.ReadDir(unitDefsFS, "catalog/units/"+factionKey)
		if err != nil {
			panic("catalog/units/" + factionKey + ": " + err.Error())
		}
		for _, entry := range unitEntries {
			if !entry.IsDir() {
				panic("catalog/units/" + factionKey + ": unexpected file " + entry.Name() + " — units must live at catalog/units/<faction>/<unit>/<unit>.json")
			}
			unitKey := entry.Name()
			rel := "catalog/units/" + factionKey + "/" + unitKey + "/" + unitKey + ".json"
			data, err := unitDefsFS.ReadFile(rel)
			if err != nil {
				panic(rel + ": " + err.Error())
			}
			var def UnitDef
			if err := json.Unmarshal(data, &def); err != nil {
				panic(rel + ": " + err.Error())
			}
			if def.Type == "" {
				panic(rel + `: missing "type" field`)
			}
			if def.Type != unitKey {
				panic(rel + ": def.Type " + def.Type + " does not match directory name " + unitKey)
			}
			if def.Faction != factionKey {
				panic(rel + `: def.Faction "` + def.Faction + `" does not match parent directory "` + factionKey + `"`)
			}
			if def.CombatProfile != "" {
				if _, ok := combatProfiles[def.CombatProfile]; !ok {
					panic(rel + `: combatProfile "` + def.CombatProfile + `" is not a known profile (see combat_ai_profiles.go)`)
				}
			}
			if def.DominionPointDropChance < 0 || def.DominionPointDropChance > 1 {
				panic(rel + `: unit "` + def.Type + `": dominionPointDropChance must be in [0,1]`)
			}
			if def.DominionPointAmount < 0 {
				panic(rel + `: unit "` + def.Type + `": dominionPointAmount must be >= 0`)
			}
			for _, t := range def.TargetableTypes {
				if t != TargetClassGround && t != TargetClassFlyer {
					panic(rel + `: unit "` + def.Type + `": targetableTypes entry "` + t + `" must be one of "ground" | "flyer"`)
				}
			}
			if def.DamageType != "" && !IsValidDamageType(def.DamageType) {
				panic(rel + `: damageType "` + string(def.DamageType) + `" is not a registered damage type`)
			}
			if def.Projectile != "" {
				if _, ok := getProjectileDef(def.Projectile); !ok {
					panic(rel + `: projectile "` + def.Projectile + `" is not a registered projectile def`)
				}
			}
			if def.ProjectileScale < 0 {
				panic(rel + `: unit "` + def.Type + `": projectileScale must be >= 0 (0/omitted ⇒ client default 1×)`)
			}
			if def.ChannelLoop != nil {
				if def.ChannelLoop.Start < 0 {
					panic(rel + `: unit "` + def.Type + `": channelLoop.start must be >= 0`)
				}
				if def.ChannelLoop.End < def.ChannelLoop.Start {
					panic(rel + `: unit "` + def.Type + `": channelLoop.end must be >= channelLoop.start`)
				}
			}
			if def.MaxMana < 0 || def.ManaRegenRate < 0 {
				panic(rel + `: maxMana and manaRegenRate must be >= 0`)
			}
			if len(def.PathChances) > 0 {
				// Weights are relative and normalized at roll time, so any
				// non-negative set with a positive sum is valid. Per-key
				// path-existence is cross-checked in path_defs.go init (the
				// paths catalog is not yet loaded at this var-init stage).
				var pathWeightSum float64
				for path, weight := range def.PathChances {
					if weight < 0 {
						panic(rel + `: unit "` + def.Type + `": pathChances["` + path + `"] must be >= 0`)
					}
					pathWeightSum += weight
				}
				if pathWeightSum <= 0 {
					panic(rel + `: unit "` + def.Type + `": pathChances weights must sum to > 0`)
				}
			}
			for _, b := range def.RequiresBuildings {
				if _, ok := getBuildingDef(b); !ok {
					panic(rel + `: requiresBuildings entry "` + b +
						`" is not a registered building type`)
				}
			}
			if _, dup := result[def.Type]; dup {
				panic(rel + `: duplicate unit type "` + def.Type + `" — type ids must be globally unique across factions`)
			}
			result[def.Type] = def
		}
	}
	return result
}

func getUnitDef(unitType string) (UnitDef, bool) {
	def, ok := unitDefsByType[unitType]
	return def, ok
}

func ListUnitDefs() []UnitDef {
	defs := make([]UnitDef, 0, len(unitDefsByType))
	for _, def := range unitDefsByType {
		defs = append(defs, def)
	}
	sort.Slice(defs, func(i, j int) bool { return defs[i].Type < defs[j].Type })
	return defs
}
